File: src/test/clojure/discourje/core/validation/closed_channel_must_be_used_in_path.clj
(ns discourje.core.validation.closed-channel-must-be-used-in-path
  (:require [discourje.core.spec :as s]))

(s/defrole ::a)
(s/defrole ::b)
(s/defrole ::c)
(s/defrole ::d)

;; All closed channels are also used on that path
(s/defsession ::protocol-trivial-correct []
              (s/alt
                [(s/--> ::a ::b)
                 (s/close ::a ::b)]
                [(s/--> ::a ::c)
                 (s/close ::a ::c)]))

;; Channel a-->b and a-->c are not always used on the path.
(s/defsession ::protocol-trivial-incorrect []
              (s/--> ::a ::b)
              (s/close ::a ::b)
              (s/close ::a ::c))

; Channels closed on parallel paths, just as messages. Every path should, however, contain close and send
(s/defsession ::protocol-non-trivial-correct []
              (s/par
                (s/--> ::a ::b)
                (s/--> ::b ::a))
              (s/par
                (s/close ::a ::b)
                (s/close ::b ::a)))

; For both closed a channel exists where it is used, but also a channel where it is not used
(s/defsession ::protocol-non-trivial-incorrect []
              (s/alt
                (s/--> ::a ::b)
                (s/--> ::a ::c)
                )
              (s/close ::a ::b)
              (s/close ::a ::c))

File: src/main/java/discourje/core/validation/rules/ClosedChannelMustBeUsedInPath.java
package discourje.core.validation.rules;

import discourje.core.validation.Rule;
import discourje.core.validation.formulas.CtlFormula;
import static discourje.core.validation.formulas.CtlFormulas.AG;
import static discourje.core.validation.formulas.CtlFormulas.EP;
import static discourje.core.validation.formulas.CtlFormulas.close;
import static discourje.core.validation.formulas.CtlFormulas.implies;
import static discourje.core.validation.formulas.CtlFormulas.send;

public class ClosedChannelMustBeUsedInPath extends Rule {

    @Override
    public String createErrorDescription(String r1, String r2) {
        return String.format("A channel from %s to %s is closed, but this channel is never used in the path.", r1, r2);
    }

    @Override
    public CtlFormula createCtlFormula(String r1, String r2) {
        return AG(implies(close(r1, r2), EP(send(r1, r2))));
    }
}
